torch/csrc/jit/codegen/cuda/parser.cpp

```diff
@@ -451,6 +451,350 @@ class IrParser {
           });
     }
 
+    {
+      auto ptr_op = getOperatorForLiteral(
+          "aten::batch_norm(Tensor input, Tensor? weight, Tensor? bias, Tensor? running_mean, Tensor? running_var, bool training, float momentum, float eps, bool cudnn_enabled) -> Tensor");
+      registerParseRule(
+          ptr_op,
+          [](const Node* node,
+             std::unordered_map<size_t, CgValue>& value_map) -> void {
+            auto input = value_map[node->input(0)->unique()]->as<TensorView>();
+
+            TensorView* weight = nullptr;
+            if (!node->input(1)->type()->isSubtypeOf(
+                    static_cast<c10::TypePtr>(NoneType::get()))) {
+              weight = value_map[node->input(1)->unique()]->as<TensorView>();
+            }
+
+            TensorView* bias = nullptr;
+            if (!node->input(2)->type()->isSubtypeOf(
+                    static_cast<c10::TypePtr>(NoneType::get()))) {
+              bias = value_map[node->input(2)->unique()]->as<TensorView>();
+            }
+
+            TensorView* running_mean = nullptr;
```

Comment:
This doesn't look right. We are not updating running_mean/running_var? I don't know we support I/O tensor, so I actually don't know how to do that.

Reply:
What is I/O tensor?

Reply:
Basically running_mean should be updated as:
`running_mean = running_mean * ( 1 - momentum ) + mean * momentum`

FYI, running stats are only used during inference, our test ground should cover that as well. We should do a few iterations of training followed by a few iterations of inference and compare with reference.

Reply:
I think we'd need global memory aliasing to handle the I/O tensor. Currently we have generate new tensor for the new_running_mean. I only partially implemented batch norm in this PR.

```
new_running_mean = running_mean * ( 1 - momentum ) + mean * momentum
running_mean = new_running_mean
```